tests/test_UyuniAPIClient.py

```diff
@@ -0,0 +1,497 @@
+    """
+    with pytest.raises(SessionException):
+        client.get_host_upgrades(random.randint(800, 1500))
+
+
+def test_get_host_groups(client, config, host_id):
+    """
+    Ensure that host groups can be found
+    """
+    host_groups = client.get_host_groups(
+        host_id
+    )
+    _groups = [x['system_group_name'] for x in host_groups]
+    assert config["valid_objects"]["host"]["group"] in _groups
+
+
+def test_get_host_groups_invalid_format(client):
+    """
+    Ensure that invalid formats will be discovered
+    """
+    with pytest.raises(EmptySetException):
+        client.get_host_groups("shittyrobots")
+
+
+def test_get_host_groups_nonexistent(client):
+    """
+    Ensure that host groups for invalid hosts cannot be gathered
+    """
+    with pytest.raises(SessionException):
+        client.get_host_groups(random.randint(800, 1500))
+
+
+def test_get_host_details(client, host_id):
+    """
+    Ensure that host details can be found
+    """
+    host_details = client.get_host_details(
+        host_id
+    )
+    # check some keys
+    keys = [
+        "id",
+        "profile_name",
+        "hostname",
+        "virtualization"
+    ]
+    for key in keys:
+        assert key in host_details.keys()
+
+
+def test_get_host_details_invalid_format(client):
+    """
+    Ensure that invalid formats will be discovered
+    """
+    with pytest.raises(EmptySetException):
+        client.get_host_details("sibio.tannenbusch.mitte")
+
+
+def test_get_host_details_nonexistent(client):
+    """
+    Ensure that details for invalid hosts cannot be gathered
+    """
+    with pytest.raises(SessionException):
+        client.get_host_details(random.randint(800, 1500))
+
+
+def test_get_host_actions(client, host_id):
+    """
+    Ensure that host actions can be found
+    """
+    assert client.get_host_actions(
+        host_id
+    )
+
+
+def test_get_host_actions_invalid_format(client):
+    """
+    Ensure that invalid formats will be discovered
+    """
+    with pytest.raises(EmptySetException):
+        client.get_host_actions("t1000.cyberdyne.sys")
+
+
+def test_get_host_actions_nonexistent(client):
+    """
+    Ensure that actions for invalid hosts cannot be gathered
+    """
+    with pytest.raises(SessionException):
+        client.get_host_actions(random.randint(800, 1500))
+
+
+def test_host_patch_do_install(client, host_id):
+    """
+    Ensure that patches can be installed
+    """
+    # get available patches
+    patches = client.get_host_patches(
+        host_id
+    )
+    if patches:
+        # install random patch
+        _patches = [x["id"] for x in patches]
+        action_id = client.install_patches(
+            host_id,
+            [random.choice(_patches)]
+        )
+        # wait until task finished before continuing
+        while not task_completed(
+                client.get_host_action(host_id, action_id[0])[0]
+        ):
+            # task not finished
+            time.sleep(10)
```

Comment:
I think the indentation looks weird. 

Hopefully this makes it look better and maybe a little bit better to understand.

```suggestion
        action_id = action_id[0]
        task = client.get_host_action(host_id, action_id)[0]
        while not task_completed(task):
            time.sleep(10)
            task = client.get_host_action(host_id, action_id)[0]
```